File: backend/src/main/java/domain/core/access/AccessFlags.java
package domain.core.access;

/**
 * Унифицированная флаговая шкала прав.
 *
 * <p>Любая комбинация прав = OR битов; любая проверка = AND с маской.
 * <p>Импликация флагов раскрывается через {@link #expand(int)}:
 * <ul>
 *   <li>{@code ROOT_WRITE}  ⇒ ROOT_READ | ADMIN_WRITE | ADMIN_READ | WRITE | READ</li>
 *   <li>{@code ROOT_READ}   ⇒ ADMIN_READ | READ</li>
 *   <li>{@code ADMIN_WRITE} ⇒ ADMIN_READ | WRITE | READ</li>
 *   <li>{@code ADMIN_READ}  ⇒ READ</li>
 * </ul>
 *
 * <p><b>Расщепление WRITE на INSERT/UPDATE.</b> {@link #WRITE_INSERT} (bit 1) — создание
 * новых записей (PreInsert); {@link #WRITE_UPDATE} (bit 6) — изменение/удаление существующих
 * (PreUpdate, PreDelete). {@link #WRITE} = их объединение. Раздельные биты моделируют права
 * «только добавление» / «только редактирование» через {@code AccessMetric.typeFlags}.
 */
public final class AccessFlags {

    private AccessFlags() {}

    public static final int NONE         = 0;

    /** Plain READ — подчиняется всем фильтрам и field-level правилам. */
    public static final int READ         = 1 << 0;   // 0x01

    /** Право создавать новые экземпляры агрегата (insert). */
    public static final int WRITE_INSERT = 1 << 1;   // 0x02

    /** ADMIN_READ — обходит instance-level (Hibernate filters, ownAccess); не обходит field-level. */
    public static final int ADMIN_READ   = 1 << 2;   // 0x04
    /** ADMIN_WRITE — обходит instance-level; неявно включает оба WRITE-бита. */
    public static final int ADMIN_WRITE  = 1 << 3;   // 0x08

    /** ROOT_READ — дополнительно обходит field-level HIDDEN/WRITE_ONLY. */
    public static final int ROOT_READ    = 1 << 4;   // 0x10
    /** ROOT_WRITE — обходит вообще всё, включая INIT_ONCE на непустом значении. */
    public static final int ROOT_WRITE   = 1 << 5;   // 0x20

    /** Право изменять или удалять существующие экземпляры (update/delete). */
    public static final int WRITE_UPDATE = 1 << 6;   // 0x40

    /**
     * Совмещённый ярлык «WRITE как операция». Эквивалент {@code WRITE_INSERT | WRITE_UPDATE}.
     * Используется в {@link DefaultAccess#READ_WRITE}/{@code WRITE_ONLY}/{@code INIT_ONCE} —
     * на field-level различение insert/update нерелевантно, важен факт «можно писать в поле».
     * На repo-level биты должны выставляться раздельно для тонкого разграничения дозвола.
     */
    public static final int WRITE        = WRITE_INSERT | WRITE_UPDATE;   // 0x42

    /**
     * Расширяет флаги по правилу импликации. Идемпотентна:
     * {@code expand(expand(x)) == expand(x)}.
     *
     * <p>{@code WRITE_INSERT} и {@code WRITE_UPDATE} НЕ имплицируют друг друга
     * — это намеренно, чтобы выдача «только INSERT» не подразумевала «UPDATE».
     * Однако {@code ADMIN_WRITE}/{@code ROOT_WRITE} имплицируют ОБА, потому что
     * роль админа — «может всё».
     */
    public static int expand(int flags) {
        int r = flags;
        if ((r & ROOT_WRITE)  != 0) r |= ROOT_READ | ADMIN_WRITE | ADMIN_READ | WRITE | READ;
        if ((r & ROOT_READ)   != 0) r |= ADMIN_READ | READ;
        if ((r & ADMIN_WRITE) != 0) r |= ADMIN_READ | WRITE | READ;
        if ((r & ADMIN_READ)  != 0) r |= READ;
        return r;
    }

    /** Истина, если в flags есть все требуемые биты required (с учётом импликации). */
    public static boolean has(int flags, int required) {
        return (expand(flags) & required) == required;
    }
}

File: frontend/src/auth/accessFlags.ts
import type { UserDto } from "../types/api";

/**
 * <h2>Єдине дзеркало backend-логіки {@code domain.core.access.AccessFlags}.</h2>
 *
 * <p>Один модуль-дзеркало бітової логіки доступу. Узгодження з backend'ом
 * (домен-файл {@code AccessFlags.java}) робиться рівно тут, в одному місці.
 *
 * <p><b>Це лише UI-логіка.</b> Реальний enforcement — на backend'і
 * ({@code AccessResolver}); див. примітку про апроксимацію в {@code permissions.ts}.
 */

/**
 * Бітова шкала прав — синхронна з backend's {@code AccessFlags}.
 * <ul>
 *   <li>READ          = 0x01 — звичайне читання</li>
 *   <li>WRITE_INSERT  = 0x02 — створення нових записів (insert)</li>
 *   <li>ADMIN_READ    = 0x04 — обхід instance-level при читанні</li>
 *   <li>ADMIN_WRITE   = 0x08 — обхід instance-level при записі</li>
 *   <li>ROOT_READ     = 0x10 — додатково обходить field-level HIDDEN/WRITE_ONLY</li>
 *   <li>ROOT_WRITE    = 0x20 — обходить геть усе</li>
 *   <li>WRITE_UPDATE  = 0x40 — зміна/видалення наявних записів (update/delete)</li>
 *   <li>WRITE = WRITE_INSERT | WRITE_UPDATE = 0x42 (комбінований ярлик)</li>
 * </ul>
 */
export const FLAG = {
  NONE: 0x00,
  READ: 0x01,
  WRITE_INSERT: 0x02,
  ADMIN_READ: 0x04,
  ADMIN_WRITE: 0x08,
  ROOT_READ: 0x10,
  ROOT_WRITE: 0x20,
  WRITE_UPDATE: 0x40,
  /** Комбінований «WRITE як операція» — WRITE_INSERT | WRITE_UPDATE. */
  WRITE: 0x02 | 0x40,
} as const;

/**
 * Розгортає прапори за правилом імплікації — дзеркало {@code AccessFlags.expand}.
 * <pre>
 *   ROOT_WRITE  ⇒ ROOT_READ | ADMIN_WRITE | ADMIN_READ | WRITE | READ   (WRITE = 0x42)
 *   ROOT_READ   ⇒ ADMIN_READ | READ
 *   ADMIN_WRITE ⇒ ADMIN_READ | WRITE | READ                             (WRITE = 0x42)
 *   ADMIN_READ  ⇒ READ
 * </pre>
 *
 * <p><b>Важливо:</b> {@code WRITE_INSERT} і {@code WRITE_UPDATE} НЕ імплікують
 * один одного — це навмисно (видача «лише INSERT» не означає «UPDATE»). Лише
 * {@code ADMIN_WRITE}/{@code ROOT_WRITE} включають обидва. Ідемпотентна:
 * {@code expandFlags(expandFlags(x)) === expandFlags(x)}.
 */
export function expandFlags(flags: number): number {
  let r = flags;
  if (r & FLAG.ROOT_WRITE)  r |= FLAG.ROOT_READ | FLAG.ADMIN_WRITE | FLAG.ADMIN_READ | FLAG.WRITE | FLAG.READ;
  if (r & FLAG.ROOT_READ)   r |= FLAG.ADMIN_READ | FLAG.READ;
  if (r & FLAG.ADMIN_WRITE) r |= FLAG.ADMIN_READ | FLAG.WRITE | FLAG.READ;
  if (r & FLAG.ADMIN_READ)  r |= FLAG.READ;
  return r;
}

/** Істина, якщо у flags присутні ВСІ біти required (з урахуванням імплікації).
 *  Дзеркало {@code AccessFlags.has}. */
export function hasFlags(flags: number, required: number): boolean {
  return (expandFlags(flags) & required) === required;
}

/**
 * Проста перевірка admin-прав за глобальними прапорами користувача.
 * Збігається з backend-логікою: admin = присутній біт {@code ADMIN_READ}
 * (з урахуванням імплікації від ROOT-/ADMIN_WRITE через {@link expandFlags}).
 */
export function isAdmin(user: UserDto | null): boolean {
  if (!user) return false;
  return (expandFlags(user.accessGlobalFlags ?? 0) & FLAG.ADMIN_READ) !== 0;
}
